### PR TITLE
Plotting multi-frequency data

### PR DESCRIPTION
**Pull request by bemoody**
We want to be able to plot multi-frequency records, with each signal at its native frequency (issue #337).

To see the individual samples, try:
```
import wfdb
r = wfdb.rdrecord('sample-data/03700181', smooth_frames=False, sampfrom=5000, sampto=5100)
wfdb.plot_wfdb(r, time_units='seconds', sig_style=['|-'])
```

Compare to the smoothed version:
```
r = wfdb.rdrecord('sample-data/03700181', sampfrom=5000, sampto=5100)
wfdb.plot_wfdb(r, time_units='seconds', sig_style=['|-'])
```

A more dramatic example:
```
r = wfdb.rdrecord('sample-data/wave_4', smooth_frames=False, sampto=10)
wfdb.plot_wfdb(r)
```
contrasted with:
```
r = wfdb.rdrecord('sample-data/wave_4', sampto=10)
wfdb.plot_wfdb(r)
```

To see the signals synchronized with annotations stored at different frequencies, try:
```
r = wfdb.rdrecord('sample-data/03700181', smooth_frames=False, sampfrom=5000, sampto=5625)
for ann, f in (('gqrsh', 500), ('gqrsl', 125), ('sqrs', 250)):
    a = wfdb.rdann('sample-data/03700181', ann, shift_samps=True, sampfrom=(5000*f//125), sampto=(5625*f//125))
    wfdb.plot_wfdb(r, a, time_units='seconds', title='annotator %s' % ann)
```
(Yeah, the sampfrom/sampto thing is a bit annoying.)

*Things not appearing in this pull request:* I think the default should be `sharex=True`, or at least, all of the subplots should have identical x-limits by default.  `plot_wfdb` should probably grow a `sharex` parameter.  However, `sharex=True` is incompatible with multi-frequency and `time_units='samples'`, though I think this could possibly be worked-around using: https://matplotlib.org/stable/api/_as_gen/matplotlib.axis.Axis.set_major_formatter.html

### CODE REVIEW

File: wfdb/plot/plot.py [hunk 1]
```diff
@@ -8,21 +8,127 @@
...
     Subplot individual channels of signals and/or annotations.
 
     Parameters
     ----------
-    signal : 1d or 2d numpy array, optional
-        The uniformly sampled signal to be plotted. If signal.ndim is 1, it is
-        assumed to be a one channel signal. If it is 2, axes 0 and 1, must
-        represent time and channel number respectively.
+    signal : 1d or 2d numpy array or list, optional
+        The uniformly sampled signal or signals to be plotted.  If signal
```
Comment (cx1111) [inline wfdb/plot/plot.py:127]:
```suggestion
        The signal or signals to be plotted. If signal
```

File: wfdb/plot/plot.py [hunk 2]
```diff
@@ -156,10 +280,12 @@ def get_plot_dims(signal, ann_samp):
...
```
Comment (cx1111) [inline wfdb/plot/plot.py:284]: It looks like you added `signal = _expand_channels(signal)` in `plot_items` right before  `get_plot_dims` is (solely) called. Perhaps `get_plot_dims` should only accept `signal` as a list then?

File: wfdb/plot/plot.py [hunk 3]
```diff
@@ -238,16 +365,21 @@ def create_figure(n_subplots, sharex, sharey, figsize):
...
+                sampling_freq=None):
     """
     Plot signal channels.
 
     Parameters
     ----------
-    signal : ndarray
-        Tranformed expanded signal into uniform signal.
+    signal : 1d or 2d numpy array or list
+        The uniformly sampled signal or signals to be plotted.  If signal
```
Comment (cx1111) [inline wfdb/plot/plot.py:376]:
```suggestion
        The signal or signals to be plotted. If signal
```

File: wfdb/plot/plot.py [hunk 4]
```diff
@@ -302,10 +456,12 @@ def plot_annotation(ann_samp, n_annot, ann_sym, signal, n_sig, fs, time_units,
         The number of annotations contained in the dat file.
     ann_sym : list
         The values of the annotation symbol locations.
-    signal : ndarray
-        Tranformed expanded signal into uniform signal.
-    sig_len : int
-        The signal length (per channel) of the dat file.
+    signal : 1d or 2d numpy array or list
+        The uniformly sampled signal or signals to be plotted.  If signal
```
Comment (cx1111) [inline wfdb/plot/plot.py:460]:
```suggestion
        The signal or signals to be plotted. If signal
```

File: wfdb/plot/plot.py [hunk 5]
```diff
@@ -262,34 +394,56 @@ def plot_signal(signal, sig_len, n_sig, fs, time_units, sig_style, axes):
...
+                    'hours': ch_freq * 3600
+                }
+                t = np.linspace(0, ch_len-1, ch_len)
+                t /= downsample_factor[time_units]
+            tarrays[ch_len, ch_freq] = t
+
+        axes[ch].plot(t, signal[ch], sig_style[ch], zorder=3)
 
 
 def plot_annotation(ann_samp, n_annot, ann_sym, signal, n_sig, fs, time_units,
```
Comment (cx1111) [inline wfdb/plot/plot.py:445]: I forget, is this an internal helper function? If so, add leading underscore.

File: wfdb/plot/plot.py [hunk 6]
```diff
@@ -577,8 +768,9 @@ def plot_wfdb(record=None, annotation=None, plot_sym=False,
     function, while allowing direct input of WFDB objects.
 
     If the record object is input, the function will extract from it:
-      - signal values, from the `p_signal` (priority) or `d_signal` attribute
-      - sampling frequency, from the `fs` attribute
+      - signal values, from the `e_p_signal`, `e_d_signal`, `p_signal`, or
```
Comment (cx1111) [inline wfdb/plot/plot.py:771]: In a future major version, there should definitely be an enum param for this.

**Review by cx1111 — APPROVED** [review]
Plot looks good. Very useful.

**Comment by bemoody** [comment]
> I forget, is this an internal helper function? If so, add leading underscore.

Yeah, I'm never quite sure what should be treated as "public" or "private".

I would suggest that we plan to do a major version bump in the near future, and rename *all* of the internal modules/functions/methods at that time.  See also issue #339.

**Comment by cx1111** [comment]
> > In a future major version, there should definitely be an enum param for this.
> 
> Agreed but I actually like the DWIM-by-default behavior in this case. I have a Record object, just show me what's there and don't bother me with the details.

Same. The default enum value/behavior should be that.

**Pull request merged**